src/traffic_server/InkAPI.cc

```diff
@@ -6091,16 +6092,22 @@ TSHttpTxnReenable(TSHttpTxn txnp, TSEvent event)
   // which is DEDICATED, the continuation needs to be called back on a
   // REGULAR thread.
   if (eth == nullptr || eth->tt != REGULAR || !eth->is_event_type(ET_NET)) {
-    eventProcessor.schedule_imm(new TSHttpSMCallback(sm, event), ET_NET);
+    schedule = true;
   } else {
     MUTEX_TRY_LOCK(trylock, sm->mutex, eth);
     if (!trylock.is_locked()) {
-      eventProcessor.schedule_imm(new TSHttpSMCallback(sm, event), ET_NET);
+      schedule = true;
     } else {
       ink_assert(eth->is_event_type(ET_NET));
       sm->state_api_callback((int)event, nullptr);
```

Comment:
Couldn't this be on another ET_NET thread that is not the thread_affinity thread?

Reply:
Yes, this case could be another ET_NET thread.  So we fall through to the common schedule case to set the affinity of the CB class which would have affinity otherwise.  So the reschedule which we were doing in this case any will go back to the thread associated with the state machine.

Reply:
In `state_api_callback()` we start modifying the HttpSM without any locks and the affinity thread doesn't match the current thread.  That seems like an issue.  Wouldn't it be easier to do this:
```
if (eth != nullptr && sm->getThreadAffinity() == eth) {
  // we are on the same thread as the HttpSM's thread affinity
  ink_assert(eth->is_event_type(ET_NET));
  sm->state_api_callback()
} else {
  // we are not on the same thread as the HttpSM's thread affinity, lets move to it
  TSHttpSMCallback *cb = new TSHttpSMCallback(sm, event);
  cb->setThreadAffinity(sm->getThreadAffinity());
  eventProcessor.schedule_imm(cb, ET_NET);
}
```

Reply:
You mean we shouldn't try to get the HttpSM lock here as the original thread does?  I'm not sure why we wouldn't want to do that.  It seems like you agree with the main part of the fix though, which is to set the affinity thread of the newly created callback before calling eventProcessor.schedule_imm()

Reply:
It would be fine to invoke `HttpSM::state_api_callback` in this case because that happens only if the `HttpSM` lock is held. Therefore when `HttpSM::state_api_callback` starts modifying the state machine, it is doing it under lock. It might not be the same thread, but it's still locked.

On other hand, looking at your suggested change, it calls `HttpSM::state_api_callout` without necessarily having the lock (the `HttpSM` could be locked on another thread). Is that what you wanted to avoid?

Reply:
I miss the MUTEX_TRY_LOCK for some reason when I read the code.  Yes, we should still get the lock so it would look something like this.  This code has the added benefit is that we are not modifying the HttpSM on a non-affinity thread and it is cleaner:

```
if (eth != nullptr && sm->getThreadAffinity() == eth) {
  // we are on the same thread as the HttpSM's thread affinity
  MUTEX_TRY_LOCK(trylock, sm->mutex, eth);
  if (trylock.is_locked()) {
    ink_assert(eth->is_event_type(ET_NET));
   sm->state_api_callback()
   return;
  }
}

// we are not on the same thread as the HttpSM's thread affinity or couldn't get the trylock
TSHttpSMCallback *cb = new TSHttpSMCallback(sm, event);
cb->setThreadAffinity(sm->getThreadAffinity());
eventProcessor.schedule_imm(cb, ET_NET);
```

Reply:
True, we should be able to collapse the two if clauses (ET_NET thread vs not) now that the SM affinity is always set.